File: src/Scanner.Core/Checks/AccessibleNameUtilities.cs
using System.Text.RegularExpressions;

namespace Scanner.Core.Checks;

internal static class AccessibleNameUtilities
{
    private static readonly Regex LabelForRegex = new("<label[^>]*for=\"(?<id>[^\"]+)\"[^>]*>", RegexOptions.IgnoreCase | RegexOptions.Compiled);
    private static readonly Regex LabelRangeRegex = new("<label[^>]*>.*?</label>", RegexOptions.IgnoreCase | RegexOptions.Singleline | RegexOptions.Compiled);
    private static readonly Regex IdRegex = new("id\\s*=\\s*\"(?<id>[^\"]+)\"", RegexOptions.IgnoreCase | RegexOptions.Compiled);
    private static readonly Regex TagRegex = new("<[^>]+>", RegexOptions.Compiled);

    public static HashSet<string> CollectLabelForIds(string content)
        => LabelForRegex.Matches(content)
            .Select(match => match.Groups["id"].Value)
            .Where(id => !string.IsNullOrWhiteSpace(id))
            .ToHashSet(StringComparer.OrdinalIgnoreCase);

    public static List<(int Start, int End)> CollectLabelRanges(string content)
    {
        var ranges = new List<(int Start, int End)>();
        foreach (Match match in LabelRangeRegex.Matches(content))
        {
            ranges.Add((match.Index, match.Index + match.Length));
        }

        return ranges;
    }

    public static HashSet<string> CollectElementIds(string content)
        => IdRegex.Matches(content)
            .Select(match => match.Groups["id"].Value)
            .Where(id => !string.IsNullOrWhiteSpace(id))
            .ToHashSet(StringComparer.OrdinalIgnoreCase);

    public static bool IsWithinLabel(int index, IReadOnlyList<(int Start, int End)> labelRanges)
        => labelRanges.Any(range => index >= range.Start && index <= range.End);

    public static bool HasAriaLabel(string? value) => !string.IsNullOrWhiteSpace(value);

    public static bool HasValidAriaLabelledBy(string? value, HashSet<string> ids)
    {
        if (string.IsNullOrWhiteSpace(value))
        {
            return false;
        }

        var tokens = value.Split(' ', StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries);
        foreach (var token in tokens)
        {
            if (ids.Contains(token))
            {
                return true;
            }
        }

        return false;
    }

    public static bool HasTitle(string? value) => !string.IsNullOrWhiteSpace(value);

    public static bool HasLabelForId(string? id, HashSet<string> labelForIds)
        => !string.IsNullOrWhiteSpace(id) && labelForIds.Contains(id);

    public static bool HasTextContent(string content)
    {
        var text = TagRegex.Replace(content, string.Empty);
        return !string.IsNullOrWhiteSpace(text);
    }
}

File: node_engine/src/checks/AccessibleNameUtilities.js
const labelForRegex = /<label[^>]*for="([^"]+)"[^>]*>/gi;
const labelRangeRegex = /<label[^>]*>.*?<\/label>/gsi;
const idRegex = /id\s*=\s*"([^"]+)"/gi;
const tagRegex = /<[^>]+>/g;

const collectLabelForIds = (content) => {
  const ids = new Set();
  for (const match of content.matchAll(labelForRegex)) {
    const id = match[1];
    if (id && id.trim()) {
      ids.add(id.toLowerCase());
    }
  }

  return ids;
};

const collectLabelRanges = (content) => {
  const ranges = [];
  for (const match of content.matchAll(labelRangeRegex)) {
    ranges.push({ start: match.index, end: match.index + match[0].length });
  }

  return ranges;
};

const collectElementIds = (content) => {
  const ids = new Set();
  for (const match of content.matchAll(idRegex)) {
    const id = match[1];
    if (id && id.trim()) {
      ids.add(id.toLowerCase());
    }
  }

  return ids;
};

const isWithinLabel = (index, labelRanges) =>
  labelRanges.some((range) => index >= range.start && index <= range.end);

const hasAriaLabel = (value) => Boolean(value && value.trim());

const hasValidAriaLabelledBy = (value, ids) => {
  if (!value || !value.trim()) {
    return false;
  }

  return value
    .split(" ")
    .map((token) => token.trim())
    .filter(Boolean)
    .some((token) => ids.has(token.toLowerCase()));
};

const hasTitle = (value) => Boolean(value && value.trim());

const hasLabelForId = (id, labelForIds) => Boolean(id && labelForIds.has(id.toLowerCase()));

const hasTextContent = (content) => Boolean(content.replace(tagRegex, "").trim());

module.exports = {
  collectLabelForIds,
  collectLabelRanges,
  collectElementIds,
  isWithinLabel,
  hasAriaLabel,
  hasValidAriaLabelledBy,
  hasTitle,
  hasLabelForId,
  hasTextContent
};
